### ISSUE TITLE
Plaso times out with message: <zeromq_queue> main_task_queue timeout waiting for a request.

### ISSUE THREAD

**Issue by the-IMOL** [log2timeline/plaso#1012]
**Plaso version:**
Plaso 1.5.0 (Windows x64 compiled exe)

**Operating system Plaso is running on:**
Windows 7 Professional

**Source data (file being parsed):**
500GB raw DD image file containing a single NTFS volume (Windows 7 Enterprise)

**Installation method:**
Win 64-bit packaged release

**Description of problem:**
This is a repeatable issue. Plaso 1.5.0 runs without issues for approximately 15 - 30 minutes but then times out with the **[WARNING] (MainProcess) PID:xxxxx** error messages every 4-ish seconds. 

Other notes: running plaso after cd'ing into the plaso directory and writing the output/dump file to the root on another volume (which exists on a separate physical disk) where the input dd image file also exists (in its own separate directory, however).

The command issued and which repeatedly results in this error is:
_log2timeline.exe --parsers "winevtx" D:\output.dump d:\images\500gb-DT-HDD.dd_

Ran plaso with debug "-d", although nothing useful in the output (at least that I can see). Also ran it with a log-file and the output is cut and pasted below:

```
2016-09-30 08:32:47,950 \[INFO\] (MainProcess) PID:15612 <engine> Preprocessing detected platform: Windows
2016-09-30 08:32:48,525 \[INFO\] (MainProcess) PID:15612 <interface> [PreProcess] Set attribute: sysregistry to \Windows\System32\config
2016-09-30 08:32:49,519 \[INFO\] (MainProcess) PID:15612 <interface> [PreProcess] Set attribute: code_page to cp1252
2016-09-30 08:32:49,523 \[INFO\] (MainProcess) PID:15612 <interface> [PreProcess] Set attribute: time_zone_str to EST5EDT
2016-09-30 08:32:51,888 \[INFO\] (MainProcess) PID:15612 <interface> [PreProcess] Set attribute: operating_system_version to 6.1
2016-09-30 08:32:51,905 \[INFO\] (MainProcess) PID:15612 <interface> [PreProcess] Set attribute: operating_system_product to Windows 7 Enterprise
2016-09-30 08:32:52,158 \[INFO\] (MainProcess) PID:15612 <extraction_frontend> Setting timezone to: EST5EDT
2016-09-30 08:56:34,082 \[WARNING\] (MainProcess) PID:15612 <zeromq_queue> main_task_queue timeout waiting for a request.
2016-09-30 08:56:38,088 \[WARNING\] (MainProcess) PID:15612 <zeromq_queue> main_task_queue timeout waiting for a request.
2016-09-30 08:56:42,130 \[WARNING\] (MainProcess) PID:15612 <zeromq_queue> main_task_queue timeout waiting for a request.
2016-09-30 08:56:46,138 \[WARNING\] (MainProcess) PID:15612 <zeromq_queue> main_task_queue timeout waiting for a request.
```

[ ...snip... the timeouts run continuously, every 4-ish seconds]

```
2016-09-30 09:41:06,832 \[WARNING\] (MainProcess) PID:15612 <zeromq_queue> main_task_queue timeout waiting for a request.
2016-09-30 09:41:10,838 \[WARNING\] (MainProcess) PID:15612 <zeromq_queue> main_task_queue timeout waiting for a request.
2016-09-30 09:41:14,854 \[WARNING\] (MainProcess) PID:15612 <zeromq_queue> main_task_queue timeout waiting for a request.
2016-09-30 09:41:18,861 \[WARNING\] (MainProcess) PID:15612 <zeromq_queue> main_task_queue timeout waiting for a request.
2016-09-30 09:41:22,867 \[WARNING\] (MainProcess) PID:15612 <zeromq_queue> main_task_queue timeout waiting for a request.
2016-09-30 09:41:26,880 \[WARNING\] (MainProcess) PID:15612 <zeromq_queue> main_task_queue timeout waiting for a request.
2016-09-30 09:41:30,885 \[WARNING\] (MainProcess) PID:15612 <zeromq_queue> main_task_queue timeout waiting for a request.
2016-09-30 09:41:34,901 \[WARNING\] (MainProcess) PID:15612 <zeromq_queue> main_task_queue timeout waiting for a request.
2016-09-30 09:41:42,539 \[WARNING\] (MainProcess) PID:15612 <zeromq_queue> main_task_queue timeout waiting for a request.
2016-09-30 09:41:46,552 \[WARNING\] (MainProcess) PID:15612 <zeromq_queue> main_task_queue timeout waiting for a request.
2016-09-30 09:41:50,559 \[WARNING\] (MainProcess) PID:15612 <zeromq_queue> main_task_queue timeout waiting for a request.
```

...........and at this point the command shell / DOS prompt hung and has to be killed/restarted each time.

**Comment by Onager** [comment]
A few questions:

- How many workers did you have?
- How long did you wait before killing log2timeline?
- What were the workers doing when the message was logged? What status did they have?

The message here indicates that no worker has requested a task for 4 seconds. Given that you only have the the winevtx parser enabled, it's possible that all the workers are busy extracting events from evtxs, are taking a while to do so.

**Comment by Onager** [comment]
One thing we might do here is drop the message to `info` from `warn`, as this condition isn't too concerning.

**Comment by the-IMOL** [comment]
I ran the same command two times in a row. In summary, the answers to your questions:

**- How many workers did you have?**  
>> three (0, 1, and 2) of them, both times the command was run
**- How long did you wait before killing log2timeline?**
>> The 1st time it ran for 18-19 hours. The 2nd time only 2 hours or so before I thought it hung again.
**- What were the workers doing when the message was logged? What status did they have?**
>> The only feedback from plaso I received is what's seen above in the first comment. Although while monitoring the processes (I'm recalling this from my own memory), it was an endless cycle of:

_localhostname.domain:PIDxx001 --> localhostname.domain:PIDxx002 TCP send (success)
localhostname.domain:PIDxx002 --> localhostname.domain:PIDxx001 TCP receive (success)
localhostname.domain:PIDxx003 --> localhostname.domain:PIDxx004 TCP send (success)
localhostname.domain:PIDxx004 --> localhostname.domain:PIDxx003 TCP receive (success)
repeat... and repeat... and repeat...
(note - data length was usually "1", but occasionally jumped up to 700)_

At the moment, I'm running the command again along with procmon and will let it go all afternoon and overnight as needed.

**Comment by Onager** [comment]
You don't have the window view, as in #1010, with all the workers and statuses?

The TCP traffic is the worker's sending their status information back.

**Comment by the-IMOL** [comment]
It doesn't get far enough for a window view like that. It just hangs with the repeated errors of _<zeromq_queue> main_task_queue timeout waiting for a request_.

Although I re-ran it again this morning while choosing **NO volume snapshots** to parse (instead of "all", like the last time) and this time it resulted in a different error that's looks like the one from #1009 . 

So, the exact same command was issued (log2timeline.exe --parsers "winevtx" D:\output.dump d:\images\500gb-DT-HDD.dd), with the only change being **no VSS to parse** and it resulted in the different error at the bottom. I'm going to re-run it again while choosing **parse all VSS**, which may be involved with the _<zeromq_queue> main_task_queue timeout_ error.

Worker_00 (PID: 6360) - events produced: 7359 - file: TSK:/Thin/egather/local/co
llect/summary.dll - running: True
Worker_01 (PID: 7724) - events produced: 8421 - file: TSK:/Packages1/CoverVA/App
lications/UserApps/UserApps_64bit/LanguagePlugins/ICUserApps_SIPBridge_64bit.msi
 - running: True
Worker_02 (PID: 4708) - events produced: 63638 - file: TSK:/Packages1/CoverVA/Ap
plications/UserApps/UserApps_64bit/backup/ICUserApps_64bit_LANG_es.mst - running
: True
Traceback (most recent call last):
  File "tools\log2timeline.py", line 767, in <module>
  File "tools\log2timeline.py", line 753, in Main
  File "tools\log2timeline.py", line 664, in ProcessSources
  File "plaso\frontend\extraction_frontend.py", line 483, in ProcessSources
  File "plaso\multi_processing\task_engine.py", line 791, in ProcessSources
  File "plaso\multi_processing\task_engine.py", line 288, in _ProcessSources
  File "plaso\multi_processing\task_engine.py", line 417, in _ScheduleTasks
  File "plaso\multi_processing\task_engine.py", line 221, in _MergeTaskStorage
  File "plaso\storage\zip_file.py", line 3291, in StartMergeTaskStorage
  File "plaso\storage\gzip_file.py", line 305, in __init__
  File "gzip.py", line 34, in open
  File "gzip.py", line 94, in __init__
IOError: [Errno 13] Permission denied: u'tmpexoini\\merge\\c57b7020f66e4042aa14e
386dae0e740.plaso'
Failed to execute script log2timeline

**Comment by the-IMOL** [comment]
Re-ran the same command except THIS TIME with the "all" volume snapshots option selected. The command was:  _log2timeline.exe --parsers "winevtx" D:\output.dump d:\images\500gb-DT-HDD.001_ 

Once again, it resulted in the _<zeromq_queue> main_task_queue timeout waiting for a request_ error messages and the DOS window being hung. Attached at bottom are the last 400-ish lines from the procmon logging (filtered on log2timeline.exe). 

Here is the a screenscrape from the window itself. Note that both of these were produced 3 hours AFTER the final time stamps (all log2timeline.exe activity had ceased because the command line / window was hung):

2016-10-05 10:13:46,569 [WARNING] (MainProcess) PID:6364 <zeromq_queue> main_task_queue timeout waiting for a request.


......... PROCMON LOG OUTPUT FILTERED ON LOG2TIMELINE.EXE ...........

**Comment by apapadop** [comment]
I am getting the same warning messages when I attempt to kill an ongoing plaso job. It doesn't exit gracefully - in fact, it has to be killed for me to get the terminal back.

I'm processing a under-5GB E01 image of a Win7SP1x64 system I installed today - I would call this a very "light" image as it's a fresh install of Win7 with very little activity before imaging. Plaso 1.5.0 has been churning at it for >4 hours which doesn't feel right - on this hardware plaso usually does the initial pass within an hour. When I finally killed it I got:

plaso - log2timeline version 1.5.0

Source path	: /media/alex/SSD/hammer6/evidence/disk.E01
Source type	: storage media image

Identifier	PID	Status		Sources		Events		File
Main		22513	aborted		104760 (0)	86550 (0)	
Worker_00	22521	running		21815 (0)	17000 (0)	VSS1:TSK:/Windows/winsxs/x86_netfx-sbs_sys_config_install_dll_31bf3856ad364e35_6.1.7600.16385_none_bf20a9b996bb3c0e
Worker_01	22525	running		19646 (1)	14529 (4)	VSS1:TSK:/Windows/winsxs/x86_netfx-sbs_sys_enterprisesvc_dll_31bf3856ad364e35_6.1.7600.16385_none_60ffafae87253a03
Worker_02	22529	running		9920 (0)	16089 (0)	VSS1:TSK:/Windows/winsxs/x86_netfx-sbs_mscorsec_dll_31bf3856ad364e35_6.1.7600.16385_none_e5d406243d850419
Worker_03	22533	running		15559 (0)	17801 (0)	VSS1:TSK:/Windows/winsxs/x86_netfx-sbs_mscordbi_dll_31bf3856ad364e35_6.1.7600.16385_none_6306500898a352f7
Worker_04	22537	running		18234 (0)	18301 (0)	VSS1:TSK:/Windows/winsxs/x86_netfx-sbs_sys_data_dll_31bf3856ad364e35_6.1.7600.16385_none_fe6017304e1a4816
Worker_05	22541	running		11708 (0)	18298 (0)	VSS1:TSK:/Windows/winsxs/x86_netfx-sbs_iehost_dll_31bf3856ad364e35_6.1.7600.16385_none_178e80c5a3bdbcbe
Worker_06	22545	running		9411 (0)	16142 (0)	VSS1:TSK:/Windows/winsxs/x86_netfx-sbs_mscorrc_dll_31bf3856ad364e35_6.1.7600.16385_none_a1303282cea2a049

Processing aborted - waiting for clean up.

2016-10-25 18:41:10,978 [WARNING] (MainProcess) PID:22513 <engine> Terminating process: Worker_02 (PID: 22529).
2016-10-25 18:41:10,978 [WARNING] (MainProcess) PID:22513 <engine> Terminating process: Worker_03 (PID: 22533).
2016-10-25 18:41:10,978 [WARNING] (MainProcess) PID:22513 <engine> Terminating process: Worker_04 (PID: 22537).
2016-10-25 18:41:10,979 [WARNING] (MainProcess) PID:22513 <engine> Terminating process: Worker_05 (PID: 22541).
2016-10-25 18:41:10,979 [WARNING] (MainProcess) PID:22513 <engine> Terminating process: Worker_06 (PID: 22545).
2016-10-25 18:41:10,979 [WARNING] (MainProcess) PID:22513 <engine> Terminating process: Worker_00 (PID: 22521).
2016-10-25 18:41:10,979 [WARNING] (MainProcess) PID:22513 <engine> Terminating process: Worker_01 (PID: 22525).
Traceback (most recent call last):
  File "/usr/bin/log2timeline.py", line 767, in <module>
    if not Main():
  File "/usr/bin/log2timeline.py", line 753, in Main
    tool.ProcessSources()
  File "/usr/bin/log2timeline.py", line 664, in ProcessSources
    timezone=self._timezone, yara_rules_string=self._yara_rules_string)
  File "/usr/lib/python2.7/dist-packages/plaso/frontend/extraction_frontend.py", line 483, in ProcessSources
    yara_rules_string=yara_rules_string)
  File "/usr/lib/python2.7/dist-packages/plaso/multi_processing/task_engine.py", line 809, in ProcessSources
    self._StopExtractionProcesses(abort=self._abort)
  File "/usr/lib/python2.7/dist-packages/plaso/multi_processing/task_engine.py", line 555, in _StopExtractionProcesses
    self._task_queue.PushItem(plaso_queue.QueueAbort(), block=False)
  File "/usr/lib/python2.7/dist-packages/plaso/engine/zeromq_queue.py", line 762, in PushItem
    self._queue.put(item, block=False)
  File "/usr/lib/python2.7/Queue.py", line 123, in put
    raise Full
Queue.Full
2016-10-25 18:41:14,383 [WARNING] (MainProcess) PID:22513 <zeromq_queue> main_task_queue timeout waiting for a request.
2016-10-25 18:41:18,385 [WARNING] (MainProcess) PID:22513 <zeromq_queue> main_task_queue timeout waiting for a request.
<snip many more copies of last message>

The worker processes (6 of them) were zombies that had to be killed:
root     22512  0.0  0.0  86104  2492 pts/1    S+   14:13   0:00 sudo log2timeline.py hammer6.db ../evidence/disk.E01
root     22513 96.7  1.7 1053376 575388 pts/1  Sl+  14:13 261:01 /usr/bin/python /usr/bin/log2timeline.py hammer6.db ../evidence/disk.E01
root     22521 97.4  0.0      0     0 pts/1    Z+   14:14 262:36 [log2timeline.py] <defunct>
root     22525 97.5  0.0      0     0 pts/1    Z+   14:14 262:43 [log2timeline.py] <defunct>
root     22529 97.4  0.0      0     0 pts/1    Z+   14:14 262:36 [log2timeline.py] <defunct>
root     22533 97.5  0.0      0     0 pts/1    Z+   14:14 262:45 [log2timeline.py] <defunct>
root     22537 97.5  0.0      0     0 pts/1    Z+   14:14 262:41 [log2timeline.py] <defunct>
root     22541 97.4  0.0      0     0 pts/1    Z+   14:14 262:29 [log2timeline.py] <defunct>
root     22545 97.5  0.0      0     0 pts/1    Z+   14:14 262:48 [log2timeline.py] <defunct>

I don't know if this helps troubleshoot at all but I'm happy to provide the image for closer examination as it doesn't contain any customer info.

Cheers,

Alex

**Comment by Onager** [comment]
I believe this was fixed in bc346b856b8b0d70fb1285b557988c47461889f0 which resolved some queue handling issue. Apologies for not circling back on this issue earlier.
